### ISSUE TITLE
Introduce `ReevaluationHint` in QHint for a single node scheduling constraint optimization

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#128956]
/assign
/sig scheduling
/cc @kubernetes/sig-scheduling-misc @macsko @dom4ha 
 

## Summary

Introduce a new return value `ReevaluationHint` in QHint so that we don't have to re-evaluate all nodes at the scheduling retries.

## Motivation

We can roughly divide our scheduling constraints into two groups: "single node constraint" or "cross-node (or node unrelated) constraint".

For example, resource requirement is "single node constraint" because if node-a doesn't have enough CPU for pod-a, node-a is filtered out in the scheduling cycle, that is, node-a's filtering result is determined only by node-a's status.
And, pod topology spread is "cross-node constraint" because if domain-a, which node-a belongs to, has too many Pods matching topology spread, node-a is filtered out, that is, node-a's filtering result is determined by not only node-a's status, but also by other nodes' status too.

Let’s think about the requeueing scenarios of the resource fit plugin (single node constraints).
There are four registered events:
- Node/Add: Node is newly added. In this case, the Pod might be schedulable on the new node.
- Node/UpdateNodeAllocatable: Node got more resources. In this case, the Pod might be schedulable on the updated node.
- Pod/Delete: The running Pod is deleted, In this case, the Pod might be schedulable on the node where the deleted Pod was running on.
- Pod/UpdatePodScaleDown: 
  - The existing Pod is scaled down: In this case, the Pod might be schedulable on the node where the scaled-down pod is running on.
  - The unschedulable Pod itself is scaled down: In this case, the Pod might be schedulable on any nodes.

In all cases, we can determine which node(s) to re-evaluate in the next scheduling cycles.

Currently, every scheduling cycle (basically) always evaluates all the nodes.
If we could reduce the number of nodes to re-evaluate, that could be a significant performance optimization for the second or later scheduling cycles.

## Goals

- Allow QHint to return which node(s) to re-evaluate, change the scheduling cycles to consider them, and improve the scheduling latency for Pods with a single node scheduling constraint.

## Non-Goals

- Cross nodes scheduling constraint is out-of-scope for now. But, see `Future ideas` section at the bottom.
  - Given the cross node scheduling constraints are only PodAffinity and PodTopologySpread in the default scheduler, we can still get a huge benefit only with a single node scheduling constraint support.

#### QHint interface change

Change QHint interface to return which node(s) to re-evaluate.

```go
type ReevaluationHint struct {
  // NodeNames has the names of nodes which should be re-evaluated in the next scheduling cycle.
  // Empty means all nodes.
  NodeNames sets.Set[string]
}

type QueueingHintFn func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (QueueingHint, ReevaluationHint, error)
```

#### Scheduling queue change

We need to run QHints for Pods that are already in backoffQ or activeQ,
otherwise some nodes that should be re-evaluated might be overlooked.

The queue would aggregate all ReevaluationHint(s) that are returned from QHints,
and store them in PodInfo or somewhere so that the next scheduling cycle can refer to them. 
#### Story 1

There are 5k nodes in the cluster.
Pod-a is created, but a scheduling for Pod-a fails because all Nodes don't have enough CPU for this Pod.
A moment later, one Pod on Node-a is removed and now Node-a may have enough CPU for Pod-a, thus the scheduler retries the scheduling for this Pod.
We currently (basically) re-evaluate all nodes in the next scheduling cycle.

But, after this enhancement, QHint from the resource fit plugin returns Queue for this Pod/Delete event, and also it returns Node-a in ReevaluationHint.
And, at the next scheduling cycle, 4999 Nodes are pre-filtered out by ReevaluationHint, and the scheduling cycle only re-evaluate Node-a
We can save the calculation for 4999 Nodes.

#### QHint calculation cost and possible negative impact on the scheduling throughput

As mentioned earlier, the number of QHint runs would increase because we have to run QHints for Pods in activeQ/backoffQ as well.
As we experienced recently, the latency of event handling could impact the scheduling throughput negatively, mainly because of the shared lock in the queue.

We need to carefully evaluate if this enhancement really improves the throughput with some PoC.

### Cross node scheduling constraint support

Actually, we might be able to do the same for some cross node constraints too.
We can add LabelSelector in ReevaluationHint.

```go
type ReevaluationHint struct {
  // NodeNames has the names of nodes which should be re-evaluated in the next scheduling cycle.
  // Empty means all nodes.
  NodeNames sets.Set[string]
  // LabelSelector has the selector for nodes which should be re-evaluated in the next scheduling cycle.
  // Empty means all nodes.
  LabelSelector *labels.Selector
}
```

And, for example, let’s say a Pod failed with PodTopologySpread because the spreading is like: zone-a:zone-b:zone-c = 2:3:4 (maxSkew: 1). In this case, PodTopologySpread rejects all nodes in zone-b and zone-c.

And, supposing a new matching Pod is added to zone-a, which makes zone-a:zone-b:zone-c = 3:3:4.
In this case, QHint from PodTopologySpread could return the labelselector for zone-a and zone-b.

But, the problem is that, then QHint has to know the current spreading situation (2:3:4) to return the label selector.
We currently have a internal policy where QHint shouldn’t look up resources other than the ones given from the QHint paramters (= QHint should only use pod, oldObj, and newObj to return results) so that we can prevent QHints to be too complicated and slow.
So, if we want to do this, we probably have to loosen this policy.

Those are why I don’t want to include the cross node scheduling constraints in the first step.
We should see if this idea is really worthy only with single node constraints first, and then we can discuss those stuff later.

### Concurrent scheduling cycles

In the long term, this enhancement could be a first step for the concurrent scheduling cycles, our dream.
As long as `NodeNames` is not overlapped, we can run the evaluations for multiple Pods concurrently since the scheduling results won't be conflict.

**Comment by alculquicondor** [comment]
Continuing the example with the fit plugin:
Let's imagine that we receive a Node add event, so we move a pod into the active queue with that node's name.
Now, before we are able to run a scheduling cycle, there is another Node add event.

Would we still run scoring for the 2 nodes? How would it work?

**Comment by sanposhiho** [comment]
As mentioned, we need to run QHint for pods in activeQ/backoffQ.

> We need to run QHints for Pods that are already in backoffQ or activeQ,

and, yes, we'd run filter/score for the 2 nodes.

**Comment by alculquicondor** [comment]
That would make queueing hints more CPU intensive, so it has to be benchmarked against the current approach.

It sounds like it shouldn't increase memory usage significantly (we wouldn't be keeping those events any longer, because the pods are in the active queue, not in-flight). But worth thinking about memory just in case.

**Comment by sanposhiho** [comment]
Yes, in any case, we have to evaluate with benchmark whether this idea really improves things or not.

**Comment by haosdent** [comment]
+1 for this idea

A question that may be a bit out of this topic
> In the long term, this enhancement could be a first step for the concurrent scheduling cycles, our dream.

Does the concurrent scheduling cycle here mean we have multiple scheduler instances to do concurrent cycles or a single scheduler instance with multiple go routines to do concurrent cycles? Do we have any KEP doc for this?

**Comment by sanposhiho** [comment]
> Does the concurrent scheduling cycle here mean we have multiple scheduler instances to do concurrent cycles or a single scheduler instance with multiple go routines to do concurrent cycles? 

My thought is one instance with concurrent cycles. It's a high cost to share the results of `ReevaluationHint` among multiple instances. But, in any case, it's too early to discuss it.

> Do we have any KEP doc for this?

No, the next step is benchmarking with PoC, and then we can start KEP for it.

**Comment by macsko** [comment]
I think this approach might be problematic when it comes to looping through activeQ. I suppose we will need to lock the queue while running qhints for the pods there, what will block the Pop() operation and thus the scheduling cycles.

**Comment by dom4ha** [comment]
I'm not sure if we could came up with representative benchmarks, as the results may depend a lot how we write them. BTW, do we have an analysis what are the biggest pain points in scheduling performance? I'm even not suggesting tackling them in the order of priority, but rather which changes may potentially make the biggest harm.

Regarding this proposal, I wonder whether this optimization is more focused on speeding up scheduling in case of positive or missed hints? If we tackle only positive hints case, then the hint might be just a recommendation (recommended node) with a fallback to full scan, but I bet the main purpose is to optimize missed hints.

In case second case, I wonder if we could come up with an approach in which we perform such analysis once for all (independent on the number of unschedulable Pods), rather than for each one separately.

**Comment by sanposhiho** [comment]
> I think this approach might be problematic when it comes to looping through activeQ. I suppose we will need to lock the queue while running qhints for the pods there, what will block the Pop() operation and thus the scheduling cycles.

Yeah, as mentioned in the risk section, the lock stuff could be a problem.

And, for that point, actually I have a rough idea in my mind to partly decouple QHint execution from the queue's lock, which I will make sure of the feasibility via PoC.
Essentially, QHint only needs the pod object and the event, which are irrelevant to the queue.

> I'm not sure if we could came up with representative benchmarks, as the results may depend a lot how we write them.

I think the existing scheduler-perf cannot catch the improvement from this enhancement because they're too simple.
This enhancement would benefit more like scenarios that a real cluster likely has, i.e., there are many incoming unscheduled pods while there are many pods that were running but get deleted.

> BTW, do we have an analysis what are the biggest pain points in scheduling performance?

It's a very broad and hence difficult question... there're many aspects to see the bottleneck of the scheduling.

- The scheduling cycle optimization: if we optimize something in the plugin or the framework, each scheduling cycle might get faster. One step higher layer, maybe we come up with an idea for efficient plugin executions, like `Skip` status and `PreFilterResult` in the past. Also, here we also have our dream, making the scheduling cycle concurrent.
- The preemption optimization: maybe the preemption logic could be optimized, like how we select node(s)/victims and how we handle the preemptor pods afterwards.
- The retrying optimization: given the scheduling cycle handles pods one by one, we shouldn't waste the scheduling cycle as much as possible and hence we should retry the scheduling wisely. QHint belongs to the optimization for this part.
- ...possibly more stuff that we haven't paid attention to yet.

> Regarding this proposal, I wonder whether this optimization is more focused on speeding up scheduling in case of positive or missed hints? If we tackle only positive hints case, then the hint might be just a recommendation (recommended node) with a fallback to full scan, but I bet the main purpose is to optimize missed hints.

What did you mean by "positive" and "missed"?
I don't think we need to fallback to a full scan if QHint(s) return valid `ReevaluationHint`s.

**Comment by alculquicondor** [comment]
> This enhancement would benefit more like scenarios that a real cluster likely has, i.e., there are many incoming unscheduled pods while there are many pods that were running but get deleted.

We do have scenarios like that, though. It might be worth adding a bit more nuance to them.

**Comment by dom4ha** [comment]
> What did you mean by "positive" and "missed"?

Since hints are best effort, by the positive I mean that in effect of the hint a Pod got scheduled, contrary to the missed one, which just rescheduled some number of Pods that eventually ended up Unschedulable again. We probably could gain the most optimizing the time scheduler spends in effect of missed hints as they burn multiple cycles.

> I don't think we need to fallback to a full scan as long as QHint(s) return valid ReevaluationHints.

Right, so having a fallback would save us from a need to evaluate hints for Pods in ActiveQ/BackoffQ, but it would speed up scheduling cycle only for a positive hint (fallback to all in case it was missed hint).

To speed up scheduling in case of missed hits, I was actually thinking about remembering the last scheduling cycle number in each unschedulable Pod and during scheduling, considering the recorded events passed that cycle. These events would give plugins ability to assess which Nodes it's worth to consider for scheduling. It's similar to what you proposed, but it would be performed during scheduling cycle instead of QHint function. So the difference is again, no need to evaluate Qhints for Pods waiting in Active/BackoffQ. Also memory usage would be under control independent from the number of Unschedulable Pods

**Comment by sanposhiho** [comment]
> Since hints are best effort

What do you mean by best effort? I'm still don't get what you tried to mean.
QHint must not miss any events that might make Pods schedulable. The QHint results should be conservative enough to catch all the corner cases that might make Pods schedulable.

> Right, so having a fallback would save us from a need to evaluate hints for Pods in ActiveQ/BackoffQ

So, are you suggesting another option, like:
- Not run QHints for ActiveQ/BackoffQ.
- Instead, the scheduling cycle only honors `ReevaluationHint` that the Pod gets when it was in the unschedQ and the QHint returned Queue.
- It could miss some Nodes that we potentially can schedule the Pod to, and we have to have a fallback.  

So, yeah, that also works. 
This idea vs my original idea would be "possibly need a fallback, which is inefficient" vs "need to run QHints for activeQ/backoffQ, which might cause the degradation coming from the queue's lock".
The point is the latter; whether we would really get a degradation due to the lock if we run QHints for activeQ/backoffQ. 
If we wouldn't, then I believe my original idea would be better, but else, we might have to end up with your suggestion, having a fallback and not run QHint for activeQ/backoffQ.

> I was actually thinking about remembering the last scheduling cycle number in each unschedulable Pod and during scheduling, considering the recorded events passed that cycle. 

I'm not sure evaluating all recorded events to determine which node(s) to evaluate is really faster than just evaluating all nodes, especially when it comes to a bigger cluster.
The essential advantage of this enhancement is that, we can pre-filter nodes outside the scheduling cycles. Recording all events and bringing to the scheduling cycle means we need to use the additional time, either way, in the scheduling cycles.

**Comment by dom4ha** [comment]
> ? I'm still don't get what you tried to mean.
> QHint must not miss any events that might make Pods schedulable. The QHint results should be conservat

Sorry for bad wording. I just mean, they don't guarantee scheduling. As you said, they are conservative, so scheduler may still need to perform several scheduling cycles without any scheduling (or for instance schedules 1 of X waiting Pods).

> This idea vs my original idea would be "possibly need a fallback, which is inefficient" vs "need to run QHints for activeQ/backoffQ, which might cause the degradation coming from the queue's lock".

Exactly. However this approach would not help in speeding up what I call "missed hints", as each time we have fallback. I believe it's more important problem to tackle this problem than just speeding up "positive hint" case.

> The point is the latter; whether we would really get a degradation due to the lock if we run QHints for activeQ/backoffQ.

True, we can debate about it.

> Recording all events and bringing to the scheduling cycle means we need to use the additional time, either way, in the scheduling cycles.

True. So assuming that we in both cases we need to process the same amount of events, the advantage here is that we would not need to store some of them in memory per each pod. Let's say we have 10k pods waiting and 10k events corresponding to scaling up cluster. Each of the 10k pods would potentially keep 10k events.

I'm actually thinking about two things here.
1. how to make events processing as much as possible a background task
2. how to post-process events only once, so that all plugins could benefit from it

I don't know yet how to do it, but I have one appealing example in my mind which could improve performance of the fit plugin, if we could maintain some sort of indexes. For instance nodes sorted by available memory/cpu. Filtering would simply iterate over such structure until it hits the limit. Similarly, we could have a list of nodes that have updated since scheduling cycle x. Note that these are just brainstorming ideas, rather than ready solutions.

**Comment by dom4ha** [comment]
> So, memory-wise, we only need to keep the final results that are aggregated with the historical QHint runs.

But in the worst case scenario, when all 10k events apply to all 10k pods, each pod would end up with remembering those 10k nodes. It's an extreme example, but it's in principle square complexity.

> I guess you meant... s/plugins/pods? If Yes, that might be a good idea. If No, at least, your this line inspired me to come up with this idea :)

Yes and no. We can either have pre-aggregations specific to a plugin, or generic, from which all plugins could benefit from (unless it's not possible in the current framework). A generic could the one that just remembers last events.

In any case, you took my suggestion right, except that we probably cannot aggregate nodes (suggested by a given plugin) forever. This is why I thought about remembering last scheduling cycle number per each unschedulable pod and process only the nodes that were suggested after that cycle.

To achieve that we could use the trick with unseen events and
- store suggested nodes interleaved with a cycle number that produced an unschedulable node.
- when scheduler reevaluates unschedulable pod, it would iterate the table backward until it hits the cycle number it was previously looking at for that pod (let's call it a separator)
- remove this separator and if it was the first one, it would clean up the list until reaching the next separator.
I bet it might be more complicated when we get into details, though.

Yes, it's less efficient, but in this way we could keep the memory and cpu overhead under control. I suspect that in real life we'd often see sets of unschedulable pods of the same kind rather than many different waiting for different triggering events. So similarly to the QHints, we'd optimize for efficiency rather than accuracy and in combination with QHints, it should be pretty accurate and save scheduler from looking at nodes that haven't changed.

> So, it ends up, again, whether we would really get a degradation due to the lock if we go with the original proposal.

I'm mainly worry about memory overhead (as stated at the beginning), unless I'm missing something.

**Comment by sanposhiho** [comment]
> But in the worst case scenario, when all 10k events apply to all 10k pods, each pod would end up with remembering those 10k nodes. 

OK, you meant the node names not events, yes you're right. If 10k node's capacity gets scaled up at the same time, all pods might get 10k node names, which is an extreme example as you said though.

@alculquicondor What do you think? So, this enhancement (if we go with the top description as is) could increase the memory usage by `{the number of unschedulable pods} * {the number of nodes} * 253 chars (the max length of the node name)` at maximum.
Should we compromise the scheduling efficiency to eliminate this concern? or can we just say users have to increase the resources as necessary?

**Comment by dom4ha** [comment]
Let's consider more likely example to 10k of unschdulable and 100 pod delete events making room for them. Even here we seem to need 253MB of additional memory.

Since GKE supports 65k nodes, having 10k of unschedulable ones is nothing unusual. If we add one zero to those numbers (for instance 1000 completed pods that happened within the backoff time) we get 2GB.

**Comment by sanposhiho** [comment]
We officially support up to 5k nodes ([ref](https://kubernetes.io/docs/setup/best-practices/cluster-large/)). 
I know GKE does more, which is awesome, but that's GKE's responsibility, and they still can just feed more memory for the scheduler to keep their support.

**Comment by dom4ha** [comment]
Scaling up along with your fleet grow is possible only when it's linear growth, but solutions with square complexity are not scalable.

**Comment by sanposhiho** [comment]
Okay, reconsidered it for the memory pov like this, what do you think?

```go
type ClusterEventWithHint struct {
...
  // ReevaluationHintFn checks events and return which node(s) the event might make accept unschedulable pods rejected by this plugin.
  ReevaluationHintFn ReevaluationHintFn
}

// * It doesn't have `pod` (the unschedulable pod) because we cannot have which node(s) to reevaluate per pod (memory increase concern).
type ReevaluationHintFn func(logger klog.Logger, oldObj, newObj interface{}) (ReevaluationHint, error)

type ReevaluationHint struct {
  // NodeNames has the names of nodes which should be re-evaluated in the next scheduling cycle.
  // Empty means all nodes.
  NodeNames sets.Set[string]
}

// ReevaluationHintsPerPlugin has the aggregated results from ReevaluationHint of a certain plugin
// to calculate which node(s) to reevaluate at the next scheduling cycles of pods rejected by a certain plugin. 
type ReevaluationHintsPerPlugin struct {
  PluginName string

  // This is puesdo and doesn't consider how to clean up nodes from here.
  // Maybe we should do `list.List` to implement a clean up in a similar way we do for `inFlightEvents`
  Nodes map[string]RevaluateNode
}

type ReevaluateNode struct {
  // CycleNumber on ReevaluateNode means when some event happens that might make this Node accept some more Pods.
  CycleNumber uint
}

type QueuedPodInfo struct {
...
  // CycleNumber means the last scheduling cycle this Pod experienced.
  CycleNumber uint
}
```

- We store `ReevaluationHintsPerPlugin`s in the queue.
- Every event invokes `ReevaluationHintFn`.
- The result from `ReevaluationHintFn` is merged into corresponding `ReevaluationHintsPerPlugin`s.
- The scheduling cycle can know which node(s) to reevaluate by:
  - 1. gets `ReevaluationHintsPerPlugin`(s) of the plugins that the pod was rejected at the previous cycle.
  - 2. gets all `ReevaluationHintsPerPlugin.Nodes` that have `CycleNumber` older than the pod's `CycleNumber` (= the number of this Pod's last scheduling cycle). Those are the nodes to reevaluate.
  - 3. remove the pod from `ReevaluationHintsPerPlugin.Pods` and 


Memory increase would be {the number of nodes} * 253 chars * {the number of plugins (it's a fixed number)} at maximum, a linear growth, which would address your concern. And, it allows the events to be handled background, not within the scheduling cycles, which would address my concern.
What do you think?

**Comment by alculquicondor** [comment]
> What do you think? So, this enhancement (if we go with the top description as is) could (if an extreme scenario happens) increase the memory usage by {the number of unschedulable pods} * {the number of nodes} * 253 chars (the max length of the node name) at maximum.
> Should we compromise the scheduling efficiency to eliminate this concern? or can we just say users have to increase the resources of the scheduler, as necessary?

At the high level, I would like to first know what the real world scenario is that we are trying to improve. Doing these calculations will potentially do more locking and increase memory usage. For memory, yes, we can potentially throw more memory into the VMs. But locking could slow down scheduling in small clusters too.

So, I would like to see a test scenario that we can measure against, even before we start any prototype.

**Comment by dom4ha** [comment]
> At the high level, I would like to first know what the real world scenario is that we are trying to improve. Doing these calculations will potentially do more locking and increase memory usage. For memory, yes, we can potentially throw more memory into the VMs. But locking could slow down scheduling in small clusters too.
> 
> So, I would like to see a test scenario that we can measure against, even before we start any prototype.

I think I described a real world scenario in https://github.com/kubernetes/kubernetes/issues/128956#issuecomment-2505985814. When we have a lot of unschedulable pods waiting for the same/similar resource, then each time this resource becomes available, all pods get rescheduled, but let's say only one/few accepted. If we could limit the number of Nodes to reconsider, such rescheduling would be much faster.

Regarding the locking, I think we still need to come up with working concept, as some synchronization might be needed.

**Comment by dom4ha** [comment]
Potentially we could even reduce backoff time in cases where rescheduing was cheap (not sure how to measure it though), but not sure whether long backoff time is considered as one of the pain points.

**Comment by sanposhiho** [comment]
> what the real world scenario is that we are trying to improve. 

One example scenario is on `User Stories` at the top. 
Generally speaking, it could enhance all re-scheduling latency of pods that don't have cross-node constraints (pod affinity, pod topology spread, etc).

> locking could slow down scheduling in small clusters too.

Yes, as discussed, locking is something another we have to deal with. But, I have some ideas. Allow me to go a bit into the implementation detail here to explain the overview.

So, as I [mentioned](https://github.com/kubernetes/kubernetes/issues/128956#issuecomment-2500880026) earlier, I've been wondering if we could decouple QHint _a bit more_ from the queue. (even before having this issue opened.)
Essentially, QHint itself only needs the pod object and the event, which are irrelevant to the queue. So, the execution doesn't have to be protected by the lock. 
If QHint (for the original requeueing purpose) wants to move Pods from the unschedQ, it obviously has to take the queue's lock eventually, which means we cannot completely decouple QHint from the queue's lock though. (that's why I said "a bit more", not completely)
But, QHints for this purpose (or a new `ReevaluationHintFn` if we go with [a second idea](https://github.com/kubernetes/kubernetes/issues/128956#issuecomment-2506150467)) don't have to take a queue's lock; it just keeps running QHints (or `ReevaluationHintFn`) with in-coming events w/ unsched pods, and aggregate the results per pod (if we go with the top idea) or per plugin (if we go with the second idea). 
The aggregation/read of the results would need another lock, Yes. But, the time the lock is taken by event handling should be much less frequent than the queue's lock is taken by the event handling (≒ current QHint) because the lock is necessary at the aggregation only when the plugin returns non-nil `ReevaluationHint`.

**Comment by sanposhiho** [comment]
> If we allow to return many Nodes per one event, then we still risk running into the same problem of square complexity (assuming we need to store them in a list).

Where is n^2, you meant? 
We store the node names in `ReevaluationHintsPerPlugin`, and we only have `ReevaluationHintsPerPlugin` per plugin. So, the memory growth is {the number of nodes} * 253 chars * {the number of plugins (= a fixed number)}.

**Comment by dom4ha** [comment]
> Where is n^2, you meant? We store the node names in `ReevaluationHintsPerPlugin`, and we only have `ReevaluationHintsPerPlugin` per plugin. So, the memory growth is {the number of nodes} * 253 chars * {the number of plugins (= a fixed number)}.

First of all, what is an example of returning more than one node per event in 'ReevaluationHint'? I assume we cannot use set or map to eliminate duplicates, as we need to put elements on a list (to enable cleanup). If each event would produce many Nodes, then we would keep multiple duplicates.

I will think more about it, as synchronization is not trivial here.

**Comment by sanposhiho** [comment]
> what is an example of returning more than one node per event in 'ReevaluationHint'?

Just thought some plugins might want to (especially in the future when supporting cross-node constraints), but yeah, we can change ReevaluationHint to have only one node, if there's any reason.

> I assume we cannot use set or map to eliminate duplicates, as we need to put elements on a list (to enable cleanup)

I see, so you're assuming we'll have to use `list.List` for a clean up, like we do for `inFlightEvents`, like I mentioned at the comment of `ReevaluationHintsPerPlugin` above.
But, that's easily workaround-able, right? like:

```go
// ReevaluationHintsPerPlugin has the aggregated results from ReevaluationHint of a certain plugin
// to calculate which node(s) to reevaluate at the next scheduling cycles of pods rejected by a certain plugin. 
type ReevaluationHintsPerPlugin struct {
  PluginName string
  
  Nodes list.List // let's use list.List to implement a clean up in a similar way to the in-flight events...

  // RegisteredNodes caches which node(s) we already put in Nodes so that we don't push a duplicated node there.
  // The node should be cleaned up from this map too, when we remove a node from Nodes.
  RegisteredNodes sets.Set[string]
}
```

**Comment by sanposhiho** [comment]
Ah, I noticed that we should track multiple events on the same node actually. 
So, we cannot simply skip the node when it's duplicated. So, instead of the above, maybe like:

```
// ReevaluationHintsPerPlugin has the aggregated results from ReevaluationHint of a certain plugin
// to calculate which node(s) to reevaluate at the next scheduling cycles of pods rejected by a certain plugin. 
type ReevaluationHintsPerPlugin struct {
  PluginName string

  // Pods are the pods that were rejected by this plugin at the previous scheduling cycle.
  // It's sorted by QueuedPodInfo.CycleNumber.
  Pods []QueuedPodInfo

  Nodes map[string]RevaluateNode
}

type ReevaluateNode struct {
  // CycleNumber on ReevaluateNode means when some event happens that might make this Node accept some more Pods.
  // CycleNumber is cleaned up when it's older than the oldest CycleNumber in Pods.
  // And, when CycleNumbers become empty, we remove ReevaluateNode from ReevaluationHintsPerPlugin.
  CycleNumbers []uint
}
```

**Comment by dom4ha** [comment]
> Ah, I noticed that we should track multiple events on the same node actually.

Exactly. 

> So, we cannot simply skip the node when it's duplicated. So, instead of the above, maybe like:

Any map operations would be slow. What about going through existing plugins and checking how new mechanism could help them and what information they would like to store (whether returning many objects is helpful and how much)?

However, before we go deeper into possible design options, we can indeed step back and think what real world scenario we want to address.
IIUC we tackle the problem of unschedulable nodes slowing down (or even starving out) scheduling of schedulable pods.
Isn't it because we process unschedulable hight priority pods as high priority? Would it harm if we treat them as some low priority ones (in the scheduling queue only)? In the worst case they would need to preempt lower priority pods, but it already can happen since those pods are waiting arbitrary time in the backoff queue.

**Comment by sanposhiho** [comment]
> what information they would like to store (whether returning many objects is helpful and how much)

Regardless of whether we make `ReevaluationHintFn` return multiple node names or a single one, we need a duplication like that.

Example:
1. pod-a is rejected by the resource fit plugin. 
2. a scheduled pod is removed from node-a, which might make node-a accept pod-a. (QHint=Queue, ReevaluationHint=node-a)
3. pod-b is rejected by the resource fit plugin. 
4. another scheduled pod is removed from node-a, which might make node-a accept pod-b. (QHint=Queue, ReevaluationHint=node-a)
5. the scheduler starts pod-a's re-scheduling. We shouldn't clean up node-a from ReevaluationHintsPerPlugin of the resource fit plugin because pod-b is also coming.

---

But, yes, I admit I was going into the implementation detail too much before discussing a higher-level topic. 
Especially, I'd like to make a conclusion on whether we should avoid n^2 memory growth problem, in the first place.
This hugely impacts how we internally design the hint function. 

I do agree, as @dom4ha keeps saying, n^2 memory growth is _better_ to be avoided. Especially, [this line](https://github.com/kubernetes/kubernetes/issues/128956#issuecomment-2505965677) is absolutely correct.
But, it's not that we're making a system that scales infinitely; we have an official cap on the resource numbers (5k nodes / 150k pods), and we can decide the memory increase would be accepted or not, assuming those numbers.

I'd like to hear the thoughts on that from other folks, especially leads (@kubernetes/sig-scheduling-leads). I know the discussion is getting long, and let me repeat the question/context shortly here.
So, if we go with the idea proposed at the top of this issue, the scheduler could get a memory increase by `{the number of unschedulable pods} * {the number of nodes that have to be re-evaluated} * 253 chars (the max length of the node name)` at maximum. 
e.g., let's say the cluster is completely packed and 1000 pods are unschedulable. Then, 1000 nodes are updated to get more allocatable resources at the same timing somehow. In this case, the scheduler has to use `1000*1000*253chars = 253mb` memory. 
This memory is released every time each pod enters the scheduling cycle.

We can mitigate this memory increase by changing the design (call it the second design), which would probably be `{the number of nodes that have to be re-evaluated} * 253 chars * {the number of registered plugins}` (for now, depending on how we could design it). However, it would become less efficient in terms of scheduling, compared to the first idea.

So, comparing two options from the scheduling efficiency perspective:
the first design can hold which nodes to re-evaluate in the next scheduling cycle per pod. That means, if a new node with 2 core is created, the scheduler would take the new node into consideration for pods with less than 2 CPU request, and would not for pods with more than 2 CPU.
The second design cannot hold which nodes to re-evaluate per pod. It can only does per plugin probably ([an example design](https://github.com/kubernetes/kubernetes/issues/128956#issuecomment-2506150467)) (again, depending on how we could design though).
In that case, if a new node with 2 core is created, the scheduler would take the new node into consideration for all pods that are rejected by the resource fit plugin, that is, regardless of how large each pod requests CPU. So, even pods with more than 2 cpus have to evaluate a new node in a next scheduling cycle.

Of course, theoretically, the second design would still bring the optimization on the re-scheduling scenarios.

So, the question is 
we can accept the memory growth and go with the most efficient option.
vs
we shouldn't accept the memory growth and compromise the efficiency to mitigate it.

**Comment by sanposhiho** [comment]
I think having a memory limitation for the first idea as a safety net is a good idea.
But, yeah, not necessarily have to be LRU. Maybe, when hitting the memory limit, we can start removing the entries for pods with more node names (of aggregated hint results) because those pods are consuming a lot of memory, and also, either way, the more node names a Pod has, the smaller a re-scheduling improvement given by a hint result would be.

**Comment by sanposhiho** [comment]
What would the reasonable upper memory limit be?
If we end up needing to set a very low upper limit, then it might be better to reconsider, instead of going with the first idea + the memory limit.

(btw, I noticed we can reduce the memory growth by using node's UIDs instead of names. 253 chars -> 36 chars per node.)

**Comment by alculquicondor** [comment]
Uhm... maybe the algorithm could be simpler: once the number of nodes in the set equals 10% (configurable), then just do a full scheduling cycle for that pod. It could maybe be based on the percentageOfNodesToScore.

**Comment by alculquicondor** [comment]
> What would the reasonable upper memory limit be?

The guiding principle should be for memory usage to be O(NlgN) on the number of Nodes or Pods (ideally linear). That would allow for the scheduler to be ready to scale whenever the supported number of nodes increases.

**Comment by haosdent** [comment]
> 1000*1000*36chars = 36mb

From a K8s user perspective, I think such memory increment is worth if most of the unschedulable Pods in the cluster could benefit from this optimization. Because I think most K8s users could give sufficient memory for K8s control plane components, especially general physical servers today has more than 2TB (What I see in other companies).

But if this is indeed a concern from the maintainer's perspective (e.g., some users would like to run K8s in edge environments), I am not sure if we could make this optimization toggleable and let the user determine whether to enable it according to the resource restrictions of their K8s running environment.

**Comment by sanposhiho** [comment]
/remove-lifecycle stale

I didn't forget this but I have been so busy that I couldn't make time for it. Will be back hopefully after the code freeze, which usually tends to be more free than usual.

**Comment by sanposhiho** [comment]
/remove-lifecycle stale

Still worth experimenting, but low priority right now

**Comment by sanposhiho** [comment]
/remove-lifecycle rotten

though I don't think we'll consider this soon. (bandwidth/priority)

**Comment by macsko** [comment]
We have noticed that in large clusters with thousands of unscheduled pods, especially autoscaled, the number of events can make queuing hints ineffective, i.e., Node/Add or Pod/Delete events are so strong and frequent that they activate many unschedulable pods. This can strenghten problems such as https://github.com/kubernetes/kubernetes/issues/135900 or cause starvation of lower priority pods. Perhaps this feature could visibly help in such unschedulable scenarios.
